src/query/changes.go

```diff
@@ -98,24 +103,28 @@ func targetChanged(s1, s2 *core.BuildState, t1, t2 *core.BuildTarget) bool {
 // sourceHash performs a partial source hash on a target to determine if it's changed.
 // This is a bit different to the one in the build package since we can't assume everything is
 // necessarily present (and for performance reasons don't want to hash *everything*).
-func sourceHash(state *core.BuildState, target *core.BuildTarget) (hash []byte, err error) {
-	defer func() {
-		if r := recover(); r != nil {
-			err = fmt.Errorf("%s", r)
-		}
-	}()
-	h := sha1.New()
+func sourceHash(state *core.BuildState, target *core.BuildTarget) ([]byte, error) {
+	var hash []byte
 	for _, tool := range target.AllTools() {
 		if _, ok := tool.Label(); ok {
 			continue // Skip in-repo tools, that will be handled via revdeps.
 		}
-		for _, path := range tool.FullPaths(state.Graph) {
-			result, err := state.PathHasher.Hash(path, false, false)
-			if err != nil {
-				return nil, err
-			}
-			h.Write(result)
+		// Tools outside the repo shouldn't change, so hashing the resolved tool path is enough.
+		hash = append(hash, toolPathHash(state, tool)...)
+	}
+	return hash, nil
+}
+
+func toolPathHash(state *core.BuildState, tool core.BuildInput) (hash []byte) {
+	defer func() {
```

Comment:
https://github.com/thought-machine/please/blob/master/src/core/build_input.go#L173

Reply:
`FullPaths` will panic if the file doesn't. Rewriting the BuildInput implementation to return errors would probably be better, but I felt it would be overkill for now. This function should also work predictably as there should only be one path returned, which the `recover` will catch if not resolved.

Reply:
Ah yeah good spot!